python/ctsm/site_and_regional/tower_site.py

```diff
@@ -0,0 +1,413 @@
+
+# Get the ctsm util tools and then the cime tools.
+_CTSM_PYTHON = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "..", "python"))
+sys.path.insert(1, _CTSM_PYTHON)
+
+# pylint: disable=wrong-import-position, import-error, unused-import, wrong-import-order
+from ctsm import add_cime_to_path
+from ctsm.path_utils import path_to_ctsm_root
+
+from CIME import build
+from CIME.case import Case
+from CIME.utils import safe_copy, expect, symlink_force
+
+logger = logging.getLogger(__name__)
+
+
+# pylint: disable=too-many-instance-attributes
+class TowerSite:
+    """
+    Parent class to NeonSite and Plumber2Site classes.
+    ...
+    Attributes
+    ----------
+    Methods
+    -------
+    """
+
+    def __init__(self, name, start_year, end_year, start_month, end_month, finidat):
+        """
+        Initializes TowerSite with the given arguments.
+        Parameters
+        ----------
+        """
+        self.name = name
+        self.start_year = int(start_year)
+        self.end_year = int(end_year)
+        self.start_month = int(start_month)
+        self.end_month = int(end_month)
+        self.cesmroot = path_to_ctsm_root()
+        self.finidat = finidat
+
+    def __str__(self):
+        """
+        Converts ingredients of the TowerSite to string for printing.
+        """
+        return "{}\n{}".format(
+            str(self.__class__),
+            "\n".join(
+                (
+                    "{} = {}".format(str(key), str(self.__dict__[key]))
+                    for key in sorted(self.__dict__)
+                )
+            ),
+        )
+
+    def build_base_case(
+        self, cesmroot, output_root, res, compset, user_mods_dirs, overwrite=False, setup_only=False
+    ):
+        """
+        Function for building a base_case to clone.
+        To spend less time on building ctsm for the neon cases,
+        all the other cases are cloned from this case
+        Args:
+        self:
+            The NeonSite object
+        base_root (str):
+            root of the base_case CIME
+        res (str):
+            base_case resolution or gridname
+        compset (str):
+            base case compset
+        overwrite (bool) :
+            Flag to overwrite the case if exists
+        """
+        print("---- building a base case -------")
+        # pylint: disable=attribute-defined-outside-init
+        self.base_case_root = output_root
+        # pylint: enable=attribute-defined-outside-init
+        if not output_root:
+            output_root = os.getcwd()
+        case_path = os.path.join(output_root, self.name)
+
+        logger.info("base_case_name : %s", self.name)
+        logger.info("user_mods_dir  : %s", user_mods_dirs[0])
+
+        if overwrite and os.path.isdir(case_path):
+            print("Removing the existing case at: {}".format(case_path))
+            shutil.rmtree(case_path)
+
+        with Case(case_path, read_only=False) as case:
+            if not os.path.isdir(case_path):
```

Comment:
Maybe this should be brought out into a separate method?

This would be a future update...

Reply:
#2413